File: dsaLeetCode/dp/fibonacci.js
// iterative


// Time Complexity: O(n)
// Space Complexity: O(n)
const fib = (n) => {
  const dp = new Array(n + 1).fill(0);

  dp[1] = 1;

  for(let i = 2; i <= n; i++) {
    dp[i] = dp[i - 1] + dp[i - 2];
  }

  console.log(dp)

  return dp[n];
};

// Time Complexity: O(2^n)
// Space Complexity: O(2^n)
const fibonacci = (n) => {
  if(n === 1) {
    return 1;
  };

  if(n === 0) {
    return 0;
  };


  return fibonacci(n - 1) + fibonacci(n - 2);
};

// Time Complexity: O(n)
// Space Complexity: O(n)

const memo = new Map();

const fibonacciWithMemo = (n) => {
  if(n === 1) {
    return 1;
  };

  if(n === 0) {
    return 0;
  };


  if(memo.has(n)) {
    return memo.get(n)
  };

  memo.set(n, fibonacciWithMemo(n - 1) + fibonacciWithMemo(n - 2));

  return memo.get(n);
};


/*
 
fibonacci(5);
  fibonacci(4) + fibonacci(3);
    fibonacci(3) + fibonacci(2)
      fibonacci(2) + fibonacci(1)
      fibonacci(1) + fibonacci(0)
    fibonacci(2) + fibonacci(1)





 
 */

File: neetCode150/dp/fibonacci.py

"""
fib(0) = 0
fib(1) = 1

fib(n) = fib(n - 1) + fib(n - 2)


"""


# Brute force approach
# Time Complexity: O(2^n)
# Space Complexity: O(2^n)
def fibBruteForce(n):
  if n <= 1:
    return n

  return fibBruteForce(n - 1) + fibBruteForce(n - 2)

print(fibBruteForce(12))


# Top to bottom DP
# Recursive with memo (cache)
# Time Complexity: O(n)
# Space Complexity: O(n)

def fibTopToBottom(n, memo):
  if n <= 1:
    return n

  if n in memo:
    return memo[n]
  
  memo[n] = fibTopToBottom(n - 1, memo) + fibTopToBottom(n - 2, memo)
  
  return memo[n]

print(fibTopToBottom(12, {}))

# Bottom to TOP DP
# Iterative solution with an array
# Time Complexity: O(n)
# Space Complexity: O(n)
def fibBottomToTop(n):
  if n <= 1:
    return n

  dp = [0, 1]
  i = 2

  while i <= n:
    temp = dp[1]

    dp[1] = dp[0] + dp[1]
    dp[0] = temp
    i += 1

  return dp[1]

print(fibBottomToTop(12))

# Bottom to TOP DP
# Iterative solution with 2 pointers
# Time Complexity: O(n)
# Space Complexity: O(1)
def fibBottomToTopII(n):
  if n <= 1:
    return n

  p1 = 0
  p2 = 1
  i = 2

  while i <= n:
    temp = p2

    p2 = p1 + p2 
    p1 = temp
    i += 1

  return p2

print(fibBottomToTopII(100))
